### PR TITLE
Updates to Provisioner Guide for Async Backing

### PR DESCRIPTION
**Pull request by BradleyOlson64**
I suspect this may be more wordy than necessary in places, but could use some assistance deciding what to trim.

Also dispensed with some confusing naming in the provisioner tests.

### CODE REVIEW

**Comment by BradleyOlson64** [comment]
Also dispensed with some confusing naming and unused code in the provisioner tests.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2705881

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2705929

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2705996

File: roadmap/implementers-guide/src/node/utility/provisioner.md [hunk 1]
```diff
@@ -121,6 +173,26 @@ The subsystem should maintain a set of handles to Block Authorship Provisioning
 
 Forward the message to the appropriate Block Authorship Provisioning Job, or discard if no appropriate job is currently active.
 
-## Block Authorship Provisioning Job
+### Block Authorship Provisioning Job
```
Comment (mrcnski) [inline roadmap/implementers-guide/src/node/utility/provisioner.md:176]: Can the whole jobs section be removed? We have an issue open to [remove references to jobs](https://github.com/paritytech/polkadot/issues/6526) but I think *some* subsystems still use this terminology.

File: roadmap/implementers-guide/src/node/utility/provisioner.md [hunk 2]
```diff
@@ -100,6 +78,80 @@ To compute bitfield availability, then:
...
+      - If `occupied_core.next_up_on_available` is `None`, then the core being vacated is unscheduled and doesn’t need to be provisioned with a candidate.
+    - Else-if `occupied_core.time_out_at == block_number`
+      - If `Some(scheduled_core) = occupied_core.next_up_on_timeout`, the core will be vacated and in need of a provisioned candidate. A candidate is requested in exactly the same way as with `CoreState::Scheduled`.
+      - Else the core being vacated is unscheduled and doesn’t need to be provisioned with a candidate
+
+
+The end result of this process is a vector of `CandidateHash`s, sorted in order of their core index.
+
+Required Path:
```
Comment (mrcnski) [inline roadmap/implementers-guide/src/node/utility/provisioner.md:115]:
```suggestion
#### Required Path
```

File: roadmap/implementers-guide/src/node/utility/provisioner.md [hunk 3]
```diff
@@ -100,6 +78,80 @@ To compute bitfield availability, then:
...
+    - If `bitfields_indicate_availability`
+      - If `Some(scheduled_core) = occupied_core.next_up_on_available`, the core will be vacated and in need of a provisioned candidate. The provisioner requests a backable candidate from [prospective parachains](../backing/prospective-parachains.md) with the core’s scheduled `para_id` and a required path with one entry. This entry corresponds to the parablock candidate previously occupying this core, which was made available and can be built upon even though it hasn’t been seen as included in a relay chain block yet. See the Required Path section below for more detail.
+      - If `occupied_core.next_up_on_available` is `None`, then the core being vacated is unscheduled and doesn’t need to be provisioned with a candidate.
+    - Else-if `occupied_core.time_out_at == block_number`
+      - If `Some(scheduled_core) = occupied_core.next_up_on_timeout`, the core will be vacated and in need of a provisioned candidate. A candidate is requested in exactly the same way as with `CoreState::Scheduled`.
+      - Else the core being vacated is unscheduled and doesn’t need to be provisioned with a candidate
+
+
```
Comment (mrcnski) [inline roadmap/implementers-guide/src/node/utility/provisioner.md:112]:
```suggestion

```

**Review by mrcnski — APPROVED** [review]
Not too familiar with this subsystem, but this made sense to me, and I think you did a great job explaining it clearly! I really liked the links out to related subsystems, that was a good idea. I just can't say how accurate the technical part of the protocol is.

File: roadmap/implementers-guide/src/node/utility/provisioner.md [hunk 4]
```diff
@@ -2,15 +2,13 @@
...
-A major feature of the provisioner: this subsystem is responsible for ensuring that parachain block candidates are sufficiently available before sending them to potential block authors.
-
 ## Provisionable Data
 
 There are several distinct types of provisionable data, but they share this property in common: all should eventually be included in a relay chain block.
 
 ### Backed Candidates
 
-The block author can choose 0 or 1 backed parachain candidates per parachain; the only constraint is that each backed candidate has the appropriate relay parent. However, the choice of a backed candidate must be the block author's; the provisioner must ensure that block authors are aware of all available [`BackedCandidate`s](../../types/backing.md#backed-candidate).
+The block author can choose 0 or 1 backed parachain candidates per parachain; the only constraint is that each backable candidate has the appropriate relay parent. However, the choice of a backed candidate must be the block author's.
```
Comment (rphmeier) [inline roadmap/implementers-guide/src/node/utility/provisioner.md:11]: Last sentence was removed, but should be amended instead. Maybe something like "The provisioner subsystem is how those block authors make this choice in practice".

File: roadmap/implementers-guide/src/node/utility/provisioner.md [hunk 5]
```diff
@@ -45,37 +43,17 @@ Block authors request the inherent data they should use for constructing the inh
...
+When a validator is selected by BABE to author a block, it becomes a block producer. The provisioner is the subsystem best suited to choosing which specific backed candidates and availability bitfields should be assembled into the block. To engage this functionality, a `ProvisionerMessage::RequestInherentData` is sent; the response is a [`ParaInherentData`](../../types/runtime.md#parainherentdata). Each relay chain block backs at most one backable parachain block candidate per parachain. Additionally no farther block candidate can be backed until the previous one either gets declared available or expired. Appropriate bitfields, as outlined in the section on [bitfield selection](#bitfield-selection), and any dispute statements should be attached as well.
```
[the remark this answers is not in the document]
Comment (rphmeier) [reply roadmap/implementers-guide/src/node/utility/provisioner.md:46]: Now that we're revisiting this section, it seems worth noting that "declared available" can happen within the same block as the subsequent candidate being backed.

File: roadmap/implementers-guide/src/node/utility/provisioner.md [hunk 6]
```diff
@@ -100,6 +78,80 @@ To compute bitfield availability, then:
...
 - Availability has a 2/3 threshold. Therefore: `3 * availability.count_ones() >= 2 * availability.len()`
 
+### Candidate Selection: Prospective Parachains Mode
+
+The state of the provisioner `PerRelayParent` tracks an important setting, `ProspectiveParachainsMode`. This setting determines which backable candidate selection method the provisioner uses.
+
+`ProspectiveParachainsMode::Disabled` - The provisioner uses its own internal legacy candidate selection. 
+`ProspectiveParachainsMode::Enabled` - The provisioner requests that [prospective parachains](../backing/prospective-parachains.md) provide selected candidates.
+
+Candidates selected with `ProspectiveParachainsMode::Enabled` are able to benefit from the increased block production time asynchronous backing allows. For this reason, once asynchronous backing has been sufficiently tested the code related to legacy candidate selection will be removed.
```
Comment (rphmeier) [inline roadmap/implementers-guide/src/node/utility/provisioner.md:88]: Not tested, but rather once the legacy code is obsolete.

File: roadmap/implementers-guide/src/node/utility/provisioner.md [hunk 7]
```diff
@@ -100,6 +78,80 @@ To compute bitfield availability, then:
...
+The goal of candidate selection is to determine which cores are free, and then to the degree possible, pick a candidate appropriate to each free core. In prospective parachains candidate selection the provisioner handles the former process while [prospective parachains](../backing/prospective-parachains.md) handles the latter.
+
+To select backable candidates:
+
+- Get the list of core states from the runtime API
+- For each core state:
+  - On `CoreState::Free`
+    - The core is unscheduled and doesn’t need to be provisioned with a candidate
+  - On `CoreState::Scheduled`
+    - The availability core is scheduled to secure availability for the next block for a particular `para_id`. Also the core is not currently occupied by a candidate pending availability.
```
Comment (rphmeier) [inline roadmap/implementers-guide/src/node/utility/provisioner.md:101]: The wording "secure availability" wording feels awkward. I suggest "The core is unoccupied and scheduled to accept a backed block for a particular `para_id`."

File: roadmap/implementers-guide/src/node/utility/provisioner.md [hunk 8]
```diff
@@ -100,6 +78,80 @@ To compute bitfield availability, then:
...
+      - Else the core being vacated is unscheduled and doesn’t need to be provisioned with a candidate
+
+
+The end result of this process is a vector of `CandidateHash`s, sorted in order of their core index.
+
+Required Path:
+
+Required path is a parameter for `ProspectiveParachainsMessage::GetBackableCandidate`, which the provisioner sends in candidate selection. 
+
+An empty required path indicates that the requested candidate should be a direct child of the most recent parablock for the given `para_id` as of the given relay parent.
```
Comment (rphmeier) [inline roadmap/implementers-guide/src/node/utility/provisioner.md:119]: let's say "most recently included"

File: roadmap/implementers-guide/src/node/utility/provisioner.md [hunk 9]
```diff
@@ -100,6 +78,80 @@ To compute bitfield availability, then:
...
+
+An empty required path indicates that the requested candidate should be a direct child of the most recent parablock for the given `para_id` as of the given relay parent. 
+
+In contrast, a required path with one or more entries prompts [prospective parachains](../backing/prospective-parachains.md) to step forward through its fragment tree for the given `para_id` and relay parent until the desired parablock is reached. We then select a direct child of that parablock to pass to the provisioner. 
+
+The parablocks making up a required path do not need to have been previously seen as included in relay chain blocks. Thus the ability to provision backable candidates based on a required path effectively decouples backing from inclusion, resulting in the tremendous performance increase of Asynchronous Backing.
+
+> TODO: The provisioner’s use of `required_path` doesn't work for parathreads, since no candidate will be provided if `scheduled_core.para_id != occupied_core.candidate_descriptor.para_id`. We lean hard on the assumption that cores are fixed to specific parachains within a session. https://github.com/paritytech/polkadot/issues/5492
```
Comment (rphmeier) [inline roadmap/implementers-guide/src/node/utility/provisioner.md:125]: Let's leave this out of the guide, as there are new mechanisms coming in to deal with this.

File: roadmap/implementers-guide/src/node/utility/provisioner.md [hunk 10]
```diff
@@ -100,6 +78,80 @@ To compute bitfield availability, then:
...
+
+Required Path:
+
+Required path is a parameter for `ProspectiveParachainsMessage::GetBackableCandidate`, which the provisioner sends in candidate selection. 
+
+An empty required path indicates that the requested candidate should be a direct child of the most recent parablock for the given `para_id` as of the given relay parent. 
+
+In contrast, a required path with one or more entries prompts [prospective parachains](../backing/prospective-parachains.md) to step forward through its fragment tree for the given `para_id` and relay parent until the desired parablock is reached. We then select a direct child of that parablock to pass to the provisioner. 
+
+The parablocks making up a required path do not need to have been previously seen as included in relay chain blocks. Thus the ability to provision backable candidates based on a required path effectively decouples backing from inclusion, resulting in the tremendous performance increase of Asynchronous Backing.
```
Comment (rphmeier) [inline roadmap/implementers-guide/src/node/utility/provisioner.md:123]: Can you omit the last clause starting with "resulting"? Feels better to leave this section 'drier'.

File: roadmap/implementers-guide/src/node/utility/provisioner.md [hunk 11]
```diff
@@ -121,6 +173,26 @@ The subsystem should maintain a set of handles to Block Authorship Provisioning
...
 
-## Block Authorship Provisioning Job
+### Block Authorship Provisioning Job
 
 Maintain the set of channels to block authors. On receiving provisionable data, send a copy over each channel.
+
+## Glossary
+
+- **Relay-parent:** 
+  - A particular relay-chain block to which a process, perspective, and/or subset of state is linked.
```
[the remark this answers is not in the document]
Comment (rphmeier) [reply roadmap/implementers-guide/src/node/utility/provisioner.md:183]: I also find this wording confusing. How about "A particular relay-chain block which serves as an anchor and reference point for processes and data which depend on relay-chain state"?

File: roadmap/implementers-guide/src/node/utility/provisioner.md (hunk 1 above)
Comment (BradleyOlson64) [reply roadmap/implementers-guide/src/node/utility/provisioner.md:176]: Seems reasonable to get rid of given the context you provided.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2715087

File: roadmap/implementers-guide/src/node/utility/provisioner.md (hunk 8 above)
Comment (BradleyOlson64) [reply roadmap/implementers-guide/src/node/utility/provisioner.md:117]: That does sound better!

File: roadmap/implementers-guide/src/node/utility/provisioner.md [hunk 12]
```diff
@@ -121,6 +169,22 @@ The subsystem should maintain a set of handles to Block Authorship Provisioning
 
 Forward the message to the appropriate Block Authorship Provisioning Job, or discard if no appropriate job is currently active.
 
-## Block Authorship Provisioning Job
-
-Maintain the set of channels to block authors. On receiving provisionable data, send a copy over each channel.
```
Comment (mrcnski) [inline roadmap/implementers-guide/src/node/utility/provisioner.md:126]: Nice. Just to make sure, you checked that the job terminology is no longer present in the code? Should we delete the above sections as well in that case?

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2715256

File: roadmap/implementers-guide/src/node/utility/provisioner.md (hunk 12 above)
Comment (BradleyOlson64) [reply roadmap/implementers-guide/src/node/utility/provisioner.md:126]: There was some good info in that section to salvage, so I moved it to the protocol section

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2715349

**Review by slumber — APPROVED** [review]

**Review by rphmeier — APPROVED** [review]

**Pull request merged**